### PR TITLE
Update GDAS INIT utility to optionally use fracoro data

### PR DESCRIPTION
**Pull request by ChunxiZhang-NOAA**
## DESCRIPTION OF CHANGES: 
Modified run_pre-v14.chgres.sh, run_v15.chgres.gfs.sh, run_v16.chgres.sh run_v14.chgres.sh, run_v15.chgres.sh and config files to make the new fracoro data as an option to be used by all run_*.chgres.sh.

## TESTS CONDUCTED: 
Testing is conducted for run_pre-v14.chgres.sh, run_v15.chgres.gfs.sh, run_v16.chgres.sh run_v14.chgres.sh, and run_v15.chgres.sh for both 'gfs' and 'gdas'. The selected testing dates are: 2022050606, 2020050606, 2019050606 and 2017050606. All tests passed. 

## DEPENDENCIES:
None

## DOCUMENTATION:
None

## ISSUE: 
Related to a global-workflow issue[#863](https://github.com/NOAA-EMC/global-workflow/issues/863).

### CODE REVIEW

**Comment by GeorgeGayno-NOAA** [comment]
@ChunxiZhang-NOAA When I difference your new script with the v16 script, I see very few differences:

```
< # Run chgres using v17 netcdf history data as input.  These history
---
> # Run chgres using v16 netcdf history data as input.  These history
58,67d57
< if  [ ${CTAR} == 'C48' ] ; then
<    OCNRES='400'
< elif [ ${CTAR} == 'C96' ] ; then
<    OCNRES='100'
< elif [ ${CTAR} == 'C192' ] ; then
<    OCNRES='050'
< elif [ ${CTAR} == 'C384' ] || [ ${CTAR} == 'C768' ] || [ ${CTAR} == 'C1152' ]; then
<    OCNRES='025'
< fi
<
75,78c65,68
<  fix_dir_target_grid="${FIX_ORO}/${CTAR}.mx${OCNRES}_frac/fix_sfc"
<  mosaic_file_target_grid="${FIX_ORO}/${CTAR}.mx${OCNRES}_frac/${CTAR}_mosaic.nc"
<  orog_dir_target_grid="${FIX_ORO}/${CTAR}.mx${OCNRES}_frac"
<  orog_files_target_grid="oro_${CTAR}.mx${OCNRES}.tile1.nc","oro_${CTAR}.mx${OCNRES}.tile2.nc","oro_${CTAR}.mx${OCNRES}.tile3.nc","oro_${CTAR}.mx${OCNRES}.tile4.nc","oro_${CTAR}.mx${OCNRES}.tile5.nc","oro_${CTAR}.mx${OCNRES}.tile6.nc"
---
>  fix_dir_target_grid="${FIX_ORO}/${CTAR}/fix_sfc"
>  mosaic_file_target_grid="${FIX_ORO}/${CTAR}/${CTAR}_mosaic.nc"
>  orog_dir_target_grid="${FIX_ORO}/${CTAR}"
>  orog_files_target_grid="${CTAR}_oro_data.tile1.nc","${CTAR}_oro_data.tile2.nc","${CTAR}_oro_data.tile3.nc","${CTAR}_oro_data.tile4.nc","${CTAR}_oro_data.tile5.nc","${CTAR}_oro_data.tile6.nc"
```
Could you add logic to the existing v16 script to handle both cases instead of adding an entire new script? Suppose someone wants to use the fractional data with older data (v15 and previous)? Could logic be added to work with all the 'run' scripts?

**Review by GeorgeGayno-NOAA — CHANGES_REQUESTED** [review]
See my latest comment.

**Comment by ChunxiZhang-NOAA** [comment]
@GeorgeGayno-NOAA If use logic to control here, it also needs logic in the workflow. Currently, I use this v17 script in global workflow: https://github.com/NOAA-EMC/global-workflow/pull/1242/files. This new fracoro data should work for both fractional and non-fractional applications.

**Comment by ChunxiZhang-NOAA** [comment]
@GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?

What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?

**Comment by ChunxiZhang-NOAA** [comment]
> > @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?
> 
> What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?

Yes, we only consider v16 and higher version. GFS v14 and v15 don't support fractional grid. And the chance to use new ORO data with GFSv14 and v15 is low.

**Comment by GeorgeGayno-NOAA** [comment]
> > > @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?
> > 
> > 
> > What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?
> 
> Yes, we only consider v16 and higher version. GFS v14 and v15 don't support fractional grid. And the chance to use new ORO data with GFSv14 and v15 is low.

Let me rephrase - I am not talking about initializing v14 or v15 runs. Rather, will you want to use archived v14 or v15 data to coldstart a v16 or higher run?

**Comment by ChunxiZhang-NOAA** [comment]
> > > > @GeorgeGayno-NOAA I added the changes as you suggested. It works with the global-workflow. Do I need to add the newly added switch "FRAC_ORO" in the config file to use UTIL in standalone mode?
> > > 
> > > 
> > > What you have is better, but it will only work if the user selects v16 data as input. Suppose someone wants to coldstart a case from 2020 **and** wants to use the fracoro data? In that case v15 data will be used as input to chgres and the `run_v15.chgres.sh` will be run. That v15 script (and the pre-v15 scripts) does not use the ORO_DIR and ORO_NAME variables. Is that your intent?
> > 
> > 
> > Yes, we only consider v16 and higher version. GFS v14 and v15 don't support fractional grid. And the chance to use new ORO data with GFSv14 and v15 is low.
> 
> Let me rephrase - I am not talking about initializing v14 or v15 runs. Rather, will you want to use archived v14 or v15 data to coldstart a v16 or higher run?

Probably yes. Does it mean run_v14.chgres.sh and run_v15.chgres.sh also need to be modified?

File: util/gdas_init/run_v16.chgres.sh [hunk 1]
```diff
@@ -59,13 +59,30 @@ rm -fr $WORKDIR
 mkdir -p $WORKDIR
 cd $WORKDIR
```
Comment (GeorgeGayno-NOAA) [inline util/gdas_init/run_v16.chgres.sh:61]: Here is what I suggest. Move this logic that sets ORO_DIR and ORO_NAME to the 'config' file. Then at the bottom of the 'config' file, add an export of ORO_DIR and ORO_NAME.  Then, those variables would be available to all the chgres run scripts.

Comment (ChunxiZhang-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: @GeorgeGayno-NOAA Sounds a good plan. I will modify it as you suggested and test the code.

**Comment by ChunxiZhang-NOAA** [comment]
@GeorgeGayno-NOAA Please review the code I modified as you suggested. I also moved the assignment of CTAR into config file. I tested for v16, v15,v14, v13, and all passed for both 'gfs' and 'gdas' runs.

**Comment by GeorgeGayno-NOAA** [comment]
> @GeorgeGayno-NOAA Please review the code I modified as you suggested. I also moved the assignment of CTAR into config file. I tested for v16, v15,v14, v13, and all passed for both 'gfs' and 'gdas' runs.

I will review. And make sure your branch is up-to-date with develop.

File: util/gdas_init/config [hunk 2]
```diff
@@ -63,6 +63,8 @@ LEVS=65
 
 CDUMP=gdas
 
+FRAC_ORO=no
```
Comment (GeorgeGayno-NOAA) [inline util/gdas_init/config]: In the comments section, please define this variable.

File: util/gdas_init/run_v16retro.chgres.sh [hunk 3]
```diff
@@ -44,10 +44,8 @@ WORKDIR=${WORKDIR:-$OUTDIR/work.${MEMBER}}
 
 if [ ${MEMBER} == 'hires' ]; then
```
Comment (GeorgeGayno-NOAA) [inline util/gdas_init/run_v16retro.chgres.sh]: Be sure to update the namelist variables as you did in the other run scripts.

File: util/gdas_init/run_v16.chgres.sh (hunk 1 above)
Comment (GeorgeGayno-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: @ChunxiZhang-NOAA I gave you some bad advice. Placing this new logic in the config file won't work for GDAS runs. Let me rework the logic and get back to you.

Comment (ChunxiZhang-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: @GeorgeGayno-NOAA OK, but I tested and it worked for GDAS runs, didn't test it for the latest UFS_UTILS version though.

Comment (ChunxiZhang-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: @GeorgeGayno-NOAA My GDAS test run logs on Hera: /scratch1/NCEPDEV/hwrf/save/Chunxi.Zhang/Fixfv3/latest/UFS_UTILS/util/gdas_init

Comment (GeorgeGayno-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: The problem for GDAS runs is that the hires and enkf members can be at different resolutions. For example, the default resolutions are C192/C96. When I tested your branch, both the hires and enkf members used C192. I think the logic for setting the orog data will need to be moved to its own script and then "sourced" from each 'run' script. I moved the following logic to its own script:

```
if [ ${MEMBER} == 'gdas' ] || [ ${MEMBER} == 'gfs' ] || [ ${MEMBER} == 'hires' ] ; then
  CTAR=${CRES_HIRES}
else
  CTAR=${CRES_ENKF}
fi

if [ "${FRAC_ORO:-"no"}" = "yes" ]; then
  if  [ ${CTAR} == 'C48' ] ; then
    OCNRES='400'
  elif [ ${CTAR} == 'C96' ] ; then
    OCNRES='100'
  elif [ ${CTAR} == 'C192' ] ; then
    OCNRES='050'
  elif [ ${CTAR} == 'C384' ] || [ ${CTAR} == 'C768' ] || [ ${CTAR} == 'C1152' ]; then
    OCNRES='025'
  fi
  ORO_DIR="${CTAR}.mx${OCNRES}_frac"
  ORO_NAME="oro_${CTAR}.mx${OCNRES}"
else
  ORO_DIR="${CTAR}"
  ORO_NAME="${CTAR}_oro_data"
fi
```

Then 'sourced' it from one of the 'run' scripts as follows:

```
 mkdir -p $OUTDIR
 mkdir -p $OUTDIR/INPUT

+source $UFS_DIR/util/gdas_init/set_fixed_files.sh
+
 cat << EOF > fort.41
```

That seemed to work. You may have a better way.

Comment (ChunxiZhang-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: That's odd. My C192/C192 tests on both Hera and wcoss2 were successful. Which platform did you use?

Comment (GeorgeGayno-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: > That's odd. My C192/C192 tests on both Hera and wcoss2 were successful. Which platform did you use?

Try a C192/C96 test.

Comment (ChunxiZhang-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: C192/C96 were successful on both Hera and wcoss2.

Comment (GeorgeGayno-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: I am looking at your Hera test - /scratch1/NCEPDEV/hwrf/save/Chunxi.Zhang/Fixfv3/latest/UFS_UTILS/util/gdas_init. You specified a hires of C192, and enkf of C96. But if you look in log.001, chgres is reading the C192 fix files:

```
- INPUT DATA FROM FV3 GAUSSIAN NETCDF FILE.
 - OPEN TARGET GRID MOSAIC FILE:
 /scratch1/NCEPDEV/hwrf/save/Chunxi.Zhang/Fixfv3/latest/UFS_UTILS/util/gdas_init
 /../../fix/orog/C192.mx050_frac/C192_mosaic.nc
 - TRACER NAME IN OUTPUT FILE sphum
```

Comment (ChunxiZhang-NOAA) [reply util/gdas_init/run_v16.chgres.sh:61]: @GeorgeGayno-NOAA I see. Because variable 'member' has values like 1, 2, 3.... in run_*.chgres.sh, it is impossible to use CDUMP in the config file to get the correct 'CTAR'.  I made changes and put the definition of 'CTAR' back to the run_*.chgres.sh files. Then source the set_fixed_files.sh in the run_*.chgres.sh. Now tests for C192/C96 passed.

File: util/gdas_init/set_fixed_files.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+if [ "${FRAC_ORO:-"no"}" = "yes" ]; then
+  if  [ ${CTAR} == 'C48' ] ; then
+    OCNRES='400'
```
Comment (GeorgeGayno-NOAA) [inline util/gdas_init/set_fixed_files.sh]: For C48, I believe OCNRES should be 500.

**Comment by GeorgeGayno-NOAA** [comment]
@ChunxiZhang-NOAA Don't forget to update `run_v16retro.chgres.sh`.

File: util/gdas_init/set_fixed_files.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (GeorgeGayno-NOAA) [inline util/gdas_init/set_fixed_files.sh:1]: Can you add a short prolog that explains what the script does?

File: util/gdas_init/set_fixed_files.sh (hunk 4 above)
Comment (ChunxiZhang-NOAA) [reply util/gdas_init/set_fixed_files.sh]: Good catch! It should be 500. I have fixed it.

File: util/gdas_init/set_fixed_files.sh (hunk 5 above)
Comment (ChunxiZhang-NOAA) [reply util/gdas_init/set_fixed_files.sh:1]: I've added it in the script as suggested.

**Comment by ChunxiZhang-NOAA** [comment]
> @ChunxiZhang-NOAA Don't forget to update `run_v16retro.chgres.sh`.

@GeorgeGayno-NOAA I've updated and tested `run_v16retro.chgres.sh` as you suggested.

**Comment by GeorgeGayno-NOAA** [comment]
@ChunxiZhang-NOAA I think this is ready for merge. Please update your branch with the latest updates from 'develop'.

**Comment by GeorgeGayno-NOAA** [comment]
@KateFriedman-NOAA This allows the user to optionally choose the new 'frac' orog data - e.g., ./orog/C768.mx025_frac.

**Comment by GeorgeGayno-NOAA** [comment]
Tested 0ec219a with v13, v14, v15, v16 and v16 retro data. Both 'FRAC_ORO' options are working.

**Review by GeorgeGayno-NOAA — APPROVED** [review]

**Comment by KateFriedman-NOAA** [comment]
FYI @aerorahul @WalterKolczynski-NOAA

**Review by KateFriedman-NOAA — APPROVED** [review]
